In [1]:
import pandas as pd
import numpy as np
import json
import re

from sklearn.feature_extraction import DictVectorizer
from sklearn.model_selection import train_test_split, cross_val_score, GridSearchCV
from sklearn.decomposition import PCA, TruncatedSVD

from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

from nltk import ngrams, everygrams

from catboost import CatBoostClassifier, Pool, cv

### Data

In [2]:
train = pd.read_json('train.json').set_index('id')

In [3]:
train

cuisine                                        ingredients
id                                                                   
10259        greek  [romaine lettuce, black olives, grape tomatoes...
25693  southern_us  [plain flour, ground pepper, salt, tomatoes, g...
20130     filipino  [eggs, pepper, salt, mayonaise, cooking oil, g...
22213       indian                [water, vegetable oil, wheat, salt]
13162       indian  [black pepper, shallots, cornflour, cayenne pe...
...            ...                                                ...
29109        irish  [light brown sugar, granulated sugar, butter, ...
11462      italian  [KRAFT Zesty Italian Dressing, purple onion, b...
2238         irish  [eggs, citrus fruit, raisins, sourdough starte...
41882      chinese  [boneless chicken skinless thigh, minced garli...
2362       mexican  [green chile, jalapeno chilies, onions, ground...

[39774 rows x 2 columns]

In [4]:
test = pd.read_json('test.json').set_index('id')

In [5]:
test

ingredients
id                                                      
18009  [baking powder, eggs, all-purpose flour, raisi...
28583  [sugar, egg yolks, corn starch, cream of tarta...
41580  [sausage links, fennel bulb, fronds, olive oil...
29752  [meat cuts, file powder, smoked sausage, okra,...
35687  [ground black pepper, salt, sausage casings, l...
...                                                  ...
30246  [large egg yolks, fresh lemon juice, sugar, bo...
36028  [hot sauce, butter, sweet potatoes, adobo sauc...
22339  [black pepper, salt, parmigiano reggiano chees...
42525  [cheddar cheese, cayenne, paprika, plum tomato...
1443   [cold water, olive oil, lime wedges, garlic cl...

[9944 rows x 1 columns]

In [6]:
with open('sample_submission.csv') as f:
    sample_submission = f.readlines()

In [7]:
len(sample_submission)

9945

In [8]:
sample_submission[:5]

['id,cuisine\n',
 '35203,italian\n',
 '17600,italian\n',
 '35200,italian\n',
 '17602,italian\n']

### Target: cuisine

In [9]:
len(train.cuisine.unique())

20

The sample is imbalanced:

In [10]:
train.cuisine.value_counts()

italian         7838
mexican         6438
southern_us     4320
indian          3003
chinese         2673
french          2646
cajun_creole    1546
thai            1539
japanese        1423
greek           1175
spanish          989
korean           830
vietnamese       825
moroccan         821
british          804
filipino         755
irish            667
jamaican         526
russian          489
brazilian        467
Name: cuisine, dtype: int64

In [11]:
train.isna().sum()

cuisine        0
ingredients    0
dtype: int64

In [12]:
train.all(axis=0)

cuisine        True
ingredients    True
dtype: bool

### Feature engineering

In [13]:
def get_ngrams(l: list, n: int):
    ngrams_list = []
    for item in l:
        grams = list(everygrams(item, max_len=n, pad_left=True, pad_right=True, left_pad_symbol='<s>', right_pad_symbol='<e>'))
        ngrams_list += grams
    ngrams_list1 = [''.join(i) for i in ngrams_list]
    return set(ngrams_list1)

In [14]:
def clean_some_symbols(ing_list: list):
    new_list = sum([re.sub(r'[^a-zâçèéíîú!&\'\.\®’™]', ' ', ing).split() for ing in ing_list], [])
    new_list = list(set(new_list))
    return new_list

In [15]:
def clean_letters_only(ing_list: list):
    new_list = sum([re.sub(r'[^a-zâçèéíîú]', ' ', ing).split() for ing in ing_list], [])
    new_list = list(set(new_list))
    return new_list

In [16]:
def extract_ngrams_some_symbols(old_df):
    df = old_df.copy()
    df['voc_some_symbols'] = df['ingredients'].apply(lambda x: clean_some_symbols(x))
    df['some_symb_ngrams'] = df['voc_some_symbols'].apply(lambda x: get_ngrams(x, 3))
    return df[['some_symb_ngrams']].to_dict(orient='records')

In [17]:
def extract_ngrams_letters_only(old_df):
    df = old_df.copy()
    df['voc_letters_only'] = df['ingredients'].apply(lambda x: clean_some_symbols(x))
    df['letters_only_ngrams'] = df['voc_letters_only'].apply(lambda x: get_ngrams(x, 3))
    return df[['letters_only_ngrams']].to_dict(orient='records')

In [18]:
ngrams_train = extract_ngrams_letters_only(train)
ngrams_test = extract_ngrams_letters_only(test)

In [19]:
vectorizer = DictVectorizer()

In [20]:
X = vectorizer.fit_transform(ngrams_train)
X_kaggle = vectorizer.transform(ngrams_test)

In [21]:
print(X.shape)
print(X_kaggle.shape)

(39774, 4277)
(9944, 4277)


In [22]:
X_df = pd.DataFrame(X.toarray(), index=train.index, columns=vectorizer.get_feature_names_out())
X_df.head()

letters_only_ngrams=!  letters_only_ngrams=!<e>  \
id                                                       
10259                    0.0                       0.0   
25693                    0.0                       0.0   
20130                    0.0                       0.0   
22213                    0.0                       0.0   
13162                    0.0                       0.0   

       letters_only_ngrams=!<e><e>  letters_only_ngrams=!®  \
id                                                           
10259                          0.0                     0.0   
25693                          0.0                     0.0   
20130                          0.0                     0.0   
22213                          0.0                     0.0   
13162                          0.0                     0.0   

       letters_only_ngrams=!®<e>  letters_only_ngrams=&  \
id                                                        
10259                        0.0                    0.0   
25693                        0.0                    0.0   
20130                        0.0                    0.0   
22213                        0.0                    0.0   
13162                        0.0                    0.0   

       letters_only_ngrams=&<e>  letters_only_ngrams=&<e><e>  \
id                                                             
10259                       0.0                          0.0   
25693                       0.0                          0.0   
20130                       0.0                          0.0   
22213                       0.0                          0.0   
13162                       0.0                          0.0   

       letters_only_ngrams=&m  letters_only_ngrams=&ms  ...  \
id                                                      ...   
10259                     0.0                      0.0  ...   
25693                     0.0                      0.0  ...   
20130                     0.0                      0.0  ...   
22213                     0.0                      0.0  ...   
13162                     0.0                      0.0  ...   

       letters_only_ngrams=îch  letters_only_ngrams=ú  letters_only_ngrams=úa  \
id                                                                              
10259                      0.0                    0.0                     0.0   
25693                      0.0                    0.0                     0.0   
20130                      0.0                    0.0                     0.0   
22213                      0.0                    0.0                     0.0   
13162                      0.0                    0.0                     0.0   

       letters_only_ngrams=úa<e>  letters_only_ngrams=’  \
id                                                        
10259                        0.0                    0.0   
25693                        0.0                    0.0   
20130                        0.0                    0.0   
22213                        0.0                    0.0   
13162                        0.0                    0.0   

       letters_only_ngrams=’s  letters_only_ngrams=’s<e>  \
id                                                         
10259                     0.0                        0.0   
25693                     0.0                        0.0   
20130                     0.0                        0.0   
22213                     0.0                        0.0   
13162                     0.0                        0.0   

       letters_only_ngrams=™  letters_only_ngrams=™<e>  \
id                                                       
10259                    0.0                       0.0   
25693                    0.0                       0.0   
20130                    0.0                       0.0   
22213                    0.0                       0.0   
13162                    0.0                       0.0   

       letters_only_ngrams=™<e><e>  
id           

In [23]:
y = train['cuisine']
y

id
10259          greek
25693    southern_us
20130       filipino
22213         indian
13162         indian
            ...     
29109          irish
11462        italian
2238           irish
41882        chinese
2362         mexican
Name: cuisine, Length: 39774, dtype: object

### CatBoost

In [24]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=6000, random_state=10)

In [25]:
train_pool = Pool(X_train, y_train)
val_pool = Pool(X_test, y_test)

In [26]:
cb = CatBoostClassifier(objective='MultiClass', eval_metric='Accuracy', task_type='GPU', random_state=10)

In [27]:
cb.fit(train_pool, eval_set=val_pool)

Learning rate set to 0.150044
0:	learn: 0.4330846	test: 0.4358333	best: 0.4358333 (0)	total: 40.2ms	remaining: 40.2s
1:	learn: 0.4595547	test: 0.4658333	best: 0.4658333 (1)	total: 80.4ms	remaining: 40.1s
2:	learn: 0.4820868	test: 0.4885000	best: 0.4885000 (2)	total: 123ms	remaining: 41s
3:	learn: 0.4897258	test: 0.4978333	best: 0.4978333 (3)	total: 167ms	remaining: 41.6s
4:	learn: 0.4952922	test: 0.5055000	best: 0.5055000 (4)	total: 206ms	remaining: 41s
5:	learn: 0.5070764	test: 0.5120000	best: 0.5120000 (5)	total: 246ms	remaining: 40.8s
6:	learn: 0.5164032	test: 0.5226667	best: 0.5226667 (6)	total: 288ms	remaining: 40.8s
7:	learn: 0.5257891	test: 0.5326667	best: 0.5326667 (7)	total: 322ms	remaining: 39.9s
8:	learn: 0.5329543	test: 0.5421667	best: 0.5421667 (8)	total: 351ms	remaining: 38.7s
9:	learn: 0.5393794	test: 0.5483333	best: 0.5483333 (9)	total: 386ms	remaining: 38.2s
10:	learn: 0.5460117	test: 0.5566667	best: 0.5566667 (10)	total: 418ms	remaining: 37.6s
11:	learn: 0.5514301	tes

94:	learn: 0.7296145	test: 0.7141667	best: 0.7141667 (94)	total: 2.83s	remaining: 27s
95:	learn: 0.7300882	test: 0.7140000	best: 0.7141667 (94)	total: 2.86s	remaining: 26.9s
96:	learn: 0.7311541	test: 0.7148333	best: 0.7148333 (96)	total: 2.88s	remaining: 26.9s
97:	learn: 0.7316871	test: 0.7141667	best: 0.7148333 (96)	total: 2.91s	remaining: 26.8s
98:	learn: 0.7325457	test: 0.7168333	best: 0.7168333 (98)	total: 2.95s	remaining: 26.8s
99:	learn: 0.7330787	test: 0.7160000	best: 0.7168333 (98)	total: 2.98s	remaining: 26.8s
100:	learn: 0.7335820	test: 0.7178333	best: 0.7178333 (100)	total: 3s	remaining: 26.7s
101:	learn: 0.7335820	test: 0.7181667	best: 0.7181667 (101)	total: 3.03s	remaining: 26.7s
102:	learn: 0.7342334	test: 0.7183333	best: 0.7183333 (102)	total: 3.06s	remaining: 26.7s
103:	learn: 0.7350329	test: 0.7188333	best: 0.7188333 (103)	total: 3.09s	remaining: 26.6s
104:	learn: 0.7357731	test: 0.7188333	best: 0.7188333 (103)	total: 3.12s	remaining: 26.6s
105:	learn: 0.7363949	test:

186:	learn: 0.7805116	test: 0.7481667	best: 0.7488333 (184)	total: 5.44s	remaining: 23.6s
187:	learn: 0.7811334	test: 0.7480000	best: 0.7488333 (184)	total: 5.47s	remaining: 23.6s
188:	learn: 0.7813999	test: 0.7485000	best: 0.7488333 (184)	total: 5.49s	remaining: 23.6s
189:	learn: 0.7813111	test: 0.7480000	best: 0.7488333 (184)	total: 5.51s	remaining: 23.5s
190:	learn: 0.7814887	test: 0.7481667	best: 0.7488333 (184)	total: 5.54s	remaining: 23.5s
191:	learn: 0.7817552	test: 0.7490000	best: 0.7490000 (191)	total: 5.56s	remaining: 23.4s
192:	learn: 0.7820217	test: 0.7488333	best: 0.7490000 (191)	total: 5.58s	remaining: 23.4s
193:	learn: 0.7826435	test: 0.7496667	best: 0.7496667 (193)	total: 5.61s	remaining: 23.3s
194:	learn: 0.7827619	test: 0.7496667	best: 0.7496667 (193)	total: 5.64s	remaining: 23.3s
195:	learn: 0.7827619	test: 0.7496667	best: 0.7496667 (193)	total: 5.67s	remaining: 23.2s
196:	learn: 0.7829691	test: 0.7493333	best: 0.7496667 (193)	total: 5.69s	remaining: 23.2s
197:	learn

279:	learn: 0.8005270	test: 0.7650000	best: 0.7650000 (279)	total: 7.85s	remaining: 20.2s
280:	learn: 0.8006455	test: 0.7646667	best: 0.7650000 (279)	total: 7.88s	remaining: 20.1s
281:	learn: 0.8008231	test: 0.7641667	best: 0.7650000 (279)	total: 7.9s	remaining: 20.1s
282:	learn: 0.8009712	test: 0.7645000	best: 0.7650000 (279)	total: 7.92s	remaining: 20.1s
283:	learn: 0.8011192	test: 0.7646667	best: 0.7650000 (279)	total: 7.94s	remaining: 20s
284:	learn: 0.8012672	test: 0.7641667	best: 0.7650000 (279)	total: 7.97s	remaining: 20s
285:	learn: 0.8012969	test: 0.7645000	best: 0.7650000 (279)	total: 7.99s	remaining: 20s
286:	learn: 0.8013265	test: 0.7643333	best: 0.7650000 (279)	total: 8.01s	remaining: 19.9s
287:	learn: 0.8017114	test: 0.7650000	best: 0.7650000 (279)	total: 8.04s	remaining: 19.9s
288:	learn: 0.8014745	test: 0.7651667	best: 0.7651667 (288)	total: 8.06s	remaining: 19.8s
289:	learn: 0.8016225	test: 0.7650000	best: 0.7651667 (288)	total: 8.08s	remaining: 19.8s
290:	learn: 0.801

372:	learn: 0.8133772	test: 0.7703333	best: 0.7706667 (338)	total: 10.1s	remaining: 17s
373:	learn: 0.8136140	test: 0.7710000	best: 0.7710000 (373)	total: 10.2s	remaining: 17s
374:	learn: 0.8138805	test: 0.7706667	best: 0.7710000 (373)	total: 10.2s	remaining: 17s
375:	learn: 0.8139397	test: 0.7708333	best: 0.7710000 (373)	total: 10.2s	remaining: 16.9s
376:	learn: 0.8141470	test: 0.7715000	best: 0.7715000 (376)	total: 10.2s	remaining: 16.9s
377:	learn: 0.8142358	test: 0.7720000	best: 0.7720000 (377)	total: 10.3s	remaining: 16.9s
378:	learn: 0.8144727	test: 0.7706667	best: 0.7720000 (377)	total: 10.3s	remaining: 16.8s
379:	learn: 0.8145023	test: 0.7711667	best: 0.7720000 (377)	total: 10.3s	remaining: 16.8s
380:	learn: 0.8145023	test: 0.7715000	best: 0.7720000 (377)	total: 10.3s	remaining: 16.8s
381:	learn: 0.8146207	test: 0.7715000	best: 0.7720000 (377)	total: 10.3s	remaining: 16.7s
382:	learn: 0.8145319	test: 0.7718333	best: 0.7720000 (377)	total: 10.4s	remaining: 16.7s
383:	learn: 0.81

469:	learn: 0.8224670	test: 0.7741667	best: 0.7748333 (463)	total: 12.5s	remaining: 14.1s
470:	learn: 0.8225558	test: 0.7738333	best: 0.7748333 (463)	total: 12.5s	remaining: 14.1s
471:	learn: 0.8226742	test: 0.7736667	best: 0.7748333 (463)	total: 12.5s	remaining: 14s
472:	learn: 0.8229111	test: 0.7745000	best: 0.7748333 (463)	total: 12.6s	remaining: 14s
473:	learn: 0.8229407	test: 0.7745000	best: 0.7748333 (463)	total: 12.6s	remaining: 14s
474:	learn: 0.8230295	test: 0.7743333	best: 0.7748333 (463)	total: 12.6s	remaining: 13.9s
475:	learn: 0.8230295	test: 0.7746667	best: 0.7748333 (463)	total: 12.6s	remaining: 13.9s
476:	learn: 0.8234145	test: 0.7741667	best: 0.7748333 (463)	total: 12.7s	remaining: 13.9s
477:	learn: 0.8235033	test: 0.7745000	best: 0.7748333 (463)	total: 12.7s	remaining: 13.9s
478:	learn: 0.8234737	test: 0.7745000	best: 0.7748333 (463)	total: 12.7s	remaining: 13.8s
479:	learn: 0.8234737	test: 0.7746667	best: 0.7748333 (463)	total: 12.7s	remaining: 13.8s
480:	learn: 0.82

566:	learn: 0.8300172	test: 0.7771667	best: 0.7781667 (522)	total: 14.8s	remaining: 11.3s
567:	learn: 0.8299580	test: 0.7771667	best: 0.7781667 (522)	total: 14.9s	remaining: 11.3s
568:	learn: 0.8299283	test: 0.7771667	best: 0.7781667 (522)	total: 14.9s	remaining: 11.3s
569:	learn: 0.8299283	test: 0.7775000	best: 0.7781667 (522)	total: 14.9s	remaining: 11.2s
570:	learn: 0.8298691	test: 0.7771667	best: 0.7781667 (522)	total: 14.9s	remaining: 11.2s
571:	learn: 0.8301060	test: 0.7776667	best: 0.7781667 (522)	total: 15s	remaining: 11.2s
572:	learn: 0.8300468	test: 0.7775000	best: 0.7781667 (522)	total: 15s	remaining: 11.2s
573:	learn: 0.8300468	test: 0.7775000	best: 0.7781667 (522)	total: 15s	remaining: 11.1s
574:	learn: 0.8301652	test: 0.7778333	best: 0.7781667 (522)	total: 15s	remaining: 11.1s
575:	learn: 0.8301652	test: 0.7778333	best: 0.7781667 (522)	total: 15s	remaining: 11.1s
576:	learn: 0.8301652	test: 0.7778333	best: 0.7781667 (522)	total: 15.1s	remaining: 11s
577:	learn: 0.8302244	

663:	learn: 0.8361165	test: 0.7793333	best: 0.7801667 (653)	total: 17.2s	remaining: 8.68s
664:	learn: 0.8362350	test: 0.7800000	best: 0.7801667 (653)	total: 17.2s	remaining: 8.66s
665:	learn: 0.8363830	test: 0.7796667	best: 0.7801667 (653)	total: 17.2s	remaining: 8.63s
666:	learn: 0.8365015	test: 0.7795000	best: 0.7801667 (653)	total: 17.2s	remaining: 8.61s
667:	learn: 0.8365015	test: 0.7795000	best: 0.7801667 (653)	total: 17.3s	remaining: 8.58s
668:	learn: 0.8366791	test: 0.7795000	best: 0.7801667 (653)	total: 17.3s	remaining: 8.55s
669:	learn: 0.8365903	test: 0.7795000	best: 0.7801667 (653)	total: 17.3s	remaining: 8.53s
670:	learn: 0.8366495	test: 0.7793333	best: 0.7801667 (653)	total: 17.3s	remaining: 8.5s
671:	learn: 0.8367975	test: 0.7788333	best: 0.7801667 (653)	total: 17.4s	remaining: 8.47s
672:	learn: 0.8366495	test: 0.7791667	best: 0.7801667 (653)	total: 17.4s	remaining: 8.45s
673:	learn: 0.8369752	test: 0.7791667	best: 0.7801667 (653)	total: 17.4s	remaining: 8.42s
674:	learn:

762:	learn: 0.8414757	test: 0.7810000	best: 0.7823333 (739)	total: 19.5s	remaining: 6.07s
763:	learn: 0.8415349	test: 0.7813333	best: 0.7823333 (739)	total: 19.6s	remaining: 6.04s
764:	learn: 0.8414757	test: 0.7815000	best: 0.7823333 (739)	total: 19.6s	remaining: 6.02s
765:	learn: 0.8416533	test: 0.7815000	best: 0.7823333 (739)	total: 19.6s	remaining: 5.99s
766:	learn: 0.8416533	test: 0.7813333	best: 0.7823333 (739)	total: 19.6s	remaining: 5.96s
767:	learn: 0.8418902	test: 0.7815000	best: 0.7823333 (739)	total: 19.7s	remaining: 5.94s
768:	learn: 0.8420086	test: 0.7821667	best: 0.7823333 (739)	total: 19.7s	remaining: 5.91s
769:	learn: 0.8420086	test: 0.7823333	best: 0.7823333 (739)	total: 19.7s	remaining: 5.89s
770:	learn: 0.8421271	test: 0.7823333	best: 0.7823333 (739)	total: 19.7s	remaining: 5.86s
771:	learn: 0.8422455	test: 0.7825000	best: 0.7825000 (771)	total: 19.8s	remaining: 5.84s
772:	learn: 0.8424232	test: 0.7823333	best: 0.7825000 (771)	total: 19.8s	remaining: 5.81s
773:	learn

861:	learn: 0.8470421	test: 0.7856667	best: 0.7856667 (858)	total: 21.9s	remaining: 3.5s
862:	learn: 0.8471309	test: 0.7851667	best: 0.7856667 (858)	total: 21.9s	remaining: 3.48s
863:	learn: 0.8475751	test: 0.7853333	best: 0.7856667 (858)	total: 21.9s	remaining: 3.45s
864:	learn: 0.8475751	test: 0.7853333	best: 0.7856667 (858)	total: 21.9s	remaining: 3.42s
865:	learn: 0.8476047	test: 0.7851667	best: 0.7856667 (858)	total: 22s	remaining: 3.4s
866:	learn: 0.8476343	test: 0.7853333	best: 0.7856667 (858)	total: 22s	remaining: 3.37s
867:	learn: 0.8476639	test: 0.7855000	best: 0.7856667 (858)	total: 22s	remaining: 3.35s
868:	learn: 0.8477823	test: 0.7856667	best: 0.7856667 (858)	total: 22s	remaining: 3.32s
869:	learn: 0.8477527	test: 0.7858333	best: 0.7858333 (869)	total: 22.1s	remaining: 3.29s
870:	learn: 0.8477527	test: 0.7856667	best: 0.7858333 (869)	total: 22.1s	remaining: 3.27s
871:	learn: 0.8479008	test: 0.7861667	best: 0.7861667 (871)	total: 22.1s	remaining: 3.25s
872:	learn: 0.847841

958:	learn: 0.8517499	test: 0.7868333	best: 0.7868333 (958)	total: 24.2s	remaining: 1.03s
959:	learn: 0.8520163	test: 0.7865000	best: 0.7868333 (958)	total: 24.2s	remaining: 1.01s
960:	learn: 0.8520756	test: 0.7866667	best: 0.7868333 (958)	total: 24.2s	remaining: 983ms
961:	learn: 0.8520756	test: 0.7865000	best: 0.7868333 (958)	total: 24.3s	remaining: 958ms
962:	learn: 0.8520163	test: 0.7863333	best: 0.7868333 (958)	total: 24.3s	remaining: 933ms
963:	learn: 0.8523124	test: 0.7861667	best: 0.7868333 (958)	total: 24.3s	remaining: 908ms
964:	learn: 0.8522828	test: 0.7863333	best: 0.7868333 (958)	total: 24.3s	remaining: 882ms
965:	learn: 0.8524901	test: 0.7861667	best: 0.7868333 (958)	total: 24.3s	remaining: 857ms
966:	learn: 0.8524605	test: 0.7861667	best: 0.7868333 (958)	total: 24.4s	remaining: 832ms
967:	learn: 0.8525493	test: 0.7863333	best: 0.7868333 (958)	total: 24.4s	remaining: 806ms
968:	learn: 0.8527566	test: 0.7866667	best: 0.7868333 (958)	total: 24.4s	remaining: 781ms
969:	learn

In [28]:
pred_train = cb.predict(X_train)
pred_test = cb.predict(X_test)

In [29]:
print(accuracy_score(y_train, pred_train))
print(accuracy_score(y_test, pred_test))

0.8554805471664594
0.7875


Overfitting - ngrams with letters + some symbols:<br>
0.8545034642032333 (train)<br>
0.7893333333333333 (val)

Overfitting - ngrams with letters only:<br>
0.8554805471664594 (train)<br>
0.7875

In [30]:
cb = CatBoostClassifier(objective='MultiClass', task_type='GPU', random_state=10)

In [31]:
cross_val_score(cb, X, y, cv=5, scoring='accuracy').mean()

Learning rate set to 0.131472
0:	learn: 2.4799668	total: 41.8ms	remaining: 41.7s
1:	learn: 2.3062171	total: 83ms	remaining: 41.4s
2:	learn: 2.1877065	total: 129ms	remaining: 42.7s
3:	learn: 2.0906148	total: 173ms	remaining: 43s
4:	learn: 2.0142038	total: 208ms	remaining: 41.4s
5:	learn: 1.9474991	total: 243ms	remaining: 40.2s
6:	learn: 1.8889981	total: 279ms	remaining: 39.6s
7:	learn: 1.8385363	total: 313ms	remaining: 38.9s
8:	learn: 1.7952702	total: 341ms	remaining: 37.6s
9:	learn: 1.7602900	total: 370ms	remaining: 36.6s
10:	learn: 1.7211650	total: 397ms	remaining: 35.7s
11:	learn: 1.6861090	total: 431ms	remaining: 35.5s
12:	learn: 1.6616585	total: 466ms	remaining: 35.4s
13:	learn: 1.6346009	total: 498ms	remaining: 35.1s
14:	learn: 1.6133064	total: 528ms	remaining: 34.6s
15:	learn: 1.5968257	total: 552ms	remaining: 33.9s
16:	learn: 1.5784574	total: 579ms	remaining: 33.5s
17:	learn: 1.5532946	total: 612ms	remaining: 33.4s
18:	learn: 1.5414929	total: 640ms	remaining: 33s
19:	learn: 1.52

162:	learn: 0.8697936	total: 4.55s	remaining: 23.4s
163:	learn: 0.8666052	total: 4.59s	remaining: 23.4s
164:	learn: 0.8649652	total: 4.62s	remaining: 23.4s
165:	learn: 0.8633046	total: 4.64s	remaining: 23.3s
166:	learn: 0.8613418	total: 4.67s	remaining: 23.3s
167:	learn: 0.8599055	total: 4.7s	remaining: 23.3s
168:	learn: 0.8591448	total: 4.72s	remaining: 23.2s
169:	learn: 0.8576759	total: 4.75s	remaining: 23.2s
170:	learn: 0.8561811	total: 4.77s	remaining: 23.1s
171:	learn: 0.8553794	total: 4.79s	remaining: 23.1s
172:	learn: 0.8541534	total: 4.82s	remaining: 23s
173:	learn: 0.8523087	total: 4.85s	remaining: 23s
174:	learn: 0.8513006	total: 4.87s	remaining: 23s
175:	learn: 0.8495905	total: 4.9s	remaining: 22.9s
176:	learn: 0.8479888	total: 4.93s	remaining: 22.9s
177:	learn: 0.8471874	total: 4.95s	remaining: 22.9s
178:	learn: 0.8460158	total: 4.98s	remaining: 22.8s
179:	learn: 0.8446451	total: 5s	remaining: 22.8s
180:	learn: 0.8438660	total: 5.02s	remaining: 22.7s
181:	learn: 0.8423586	t

324:	learn: 0.7133554	total: 8.7s	remaining: 18.1s
325:	learn: 0.7127761	total: 8.72s	remaining: 18s
326:	learn: 0.7124864	total: 8.74s	remaining: 18s
327:	learn: 0.7119376	total: 8.76s	remaining: 18s
328:	learn: 0.7111785	total: 8.79s	remaining: 17.9s
329:	learn: 0.7105028	total: 8.81s	remaining: 17.9s
330:	learn: 0.7101463	total: 8.84s	remaining: 17.9s
331:	learn: 0.7099579	total: 8.86s	remaining: 17.8s
332:	learn: 0.7093010	total: 8.88s	remaining: 17.8s
333:	learn: 0.7080299	total: 8.91s	remaining: 17.8s
334:	learn: 0.7077100	total: 8.93s	remaining: 17.7s
335:	learn: 0.7074243	total: 8.95s	remaining: 17.7s
336:	learn: 0.7069620	total: 8.98s	remaining: 17.7s
337:	learn: 0.7063373	total: 9s	remaining: 17.6s
338:	learn: 0.7059592	total: 9.02s	remaining: 17.6s
339:	learn: 0.7055470	total: 9.05s	remaining: 17.6s
340:	learn: 0.7047758	total: 9.07s	remaining: 17.5s
341:	learn: 0.7041656	total: 9.1s	remaining: 17.5s
342:	learn: 0.7040493	total: 9.12s	remaining: 17.5s
343:	learn: 0.7037987	t

491:	learn: 0.6496365	total: 12.5s	remaining: 12.9s
492:	learn: 0.6494472	total: 12.5s	remaining: 12.9s
493:	learn: 0.6494345	total: 12.5s	remaining: 12.8s
494:	learn: 0.6494221	total: 12.6s	remaining: 12.8s
495:	learn: 0.6491320	total: 12.6s	remaining: 12.8s
496:	learn: 0.6488537	total: 12.6s	remaining: 12.8s
497:	learn: 0.6482776	total: 12.6s	remaining: 12.7s
498:	learn: 0.6480393	total: 12.6s	remaining: 12.7s
499:	learn: 0.6479555	total: 12.7s	remaining: 12.7s
500:	learn: 0.6478941	total: 12.7s	remaining: 12.6s
501:	learn: 0.6478264	total: 12.7s	remaining: 12.6s
502:	learn: 0.6471434	total: 12.7s	remaining: 12.6s
503:	learn: 0.6467589	total: 12.8s	remaining: 12.6s
504:	learn: 0.6465724	total: 12.8s	remaining: 12.5s
505:	learn: 0.6462652	total: 12.8s	remaining: 12.5s
506:	learn: 0.6460427	total: 12.8s	remaining: 12.5s
507:	learn: 0.6459605	total: 12.8s	remaining: 12.4s
508:	learn: 0.6456662	total: 12.9s	remaining: 12.4s
509:	learn: 0.6455559	total: 12.9s	remaining: 12.4s
510:	learn: 

651:	learn: 0.6090109	total: 16s	remaining: 8.56s
652:	learn: 0.6087755	total: 16.1s	remaining: 8.53s
653:	learn: 0.6085016	total: 16.1s	remaining: 8.51s
654:	learn: 0.6080928	total: 16.1s	remaining: 8.48s
655:	learn: 0.6078827	total: 16.1s	remaining: 8.46s
656:	learn: 0.6078183	total: 16.2s	remaining: 8.43s
657:	learn: 0.6077438	total: 16.2s	remaining: 8.41s
658:	learn: 0.6076700	total: 16.2s	remaining: 8.38s
659:	learn: 0.6075354	total: 16.2s	remaining: 8.35s
660:	learn: 0.6073264	total: 16.2s	remaining: 8.32s
661:	learn: 0.6072326	total: 16.3s	remaining: 8.3s
662:	learn: 0.6070863	total: 16.3s	remaining: 8.27s
663:	learn: 0.6067585	total: 16.3s	remaining: 8.24s
664:	learn: 0.6065297	total: 16.3s	remaining: 8.22s
665:	learn: 0.6065175	total: 16.3s	remaining: 8.19s
666:	learn: 0.6064752	total: 16.4s	remaining: 8.16s
667:	learn: 0.6061179	total: 16.4s	remaining: 8.14s
668:	learn: 0.6057730	total: 16.4s	remaining: 8.12s
669:	learn: 0.6057078	total: 16.4s	remaining: 8.09s
670:	learn: 0.6

812:	learn: 0.5761558	total: 19.6s	remaining: 4.5s
813:	learn: 0.5761459	total: 19.6s	remaining: 4.48s
814:	learn: 0.5760592	total: 19.6s	remaining: 4.45s
815:	learn: 0.5760233	total: 19.6s	remaining: 4.43s
816:	learn: 0.5759736	total: 19.7s	remaining: 4.4s
817:	learn: 0.5759653	total: 19.7s	remaining: 4.38s
818:	learn: 0.5759209	total: 19.7s	remaining: 4.35s
819:	learn: 0.5758146	total: 19.7s	remaining: 4.33s
820:	learn: 0.5757321	total: 19.7s	remaining: 4.3s
821:	learn: 0.5755375	total: 19.8s	remaining: 4.28s
822:	learn: 0.5754538	total: 19.8s	remaining: 4.25s
823:	learn: 0.5753142	total: 19.8s	remaining: 4.23s
824:	learn: 0.5752976	total: 19.8s	remaining: 4.2s
825:	learn: 0.5752467	total: 19.8s	remaining: 4.18s
826:	learn: 0.5750104	total: 19.9s	remaining: 4.15s
827:	learn: 0.5748952	total: 19.9s	remaining: 4.13s
828:	learn: 0.5748228	total: 19.9s	remaining: 4.1s
829:	learn: 0.5746958	total: 19.9s	remaining: 4.08s
830:	learn: 0.5745252	total: 19.9s	remaining: 4.05s
831:	learn: 0.574

974:	learn: 0.5514321	total: 23.1s	remaining: 591ms
975:	learn: 0.5511298	total: 23.1s	remaining: 568ms
976:	learn: 0.5509603	total: 23.1s	remaining: 544ms
977:	learn: 0.5507755	total: 23.1s	remaining: 520ms
978:	learn: 0.5504798	total: 23.2s	remaining: 497ms
979:	learn: 0.5502051	total: 23.2s	remaining: 473ms
980:	learn: 0.5498847	total: 23.2s	remaining: 449ms
981:	learn: 0.5496659	total: 23.2s	remaining: 426ms
982:	learn: 0.5493736	total: 23.3s	remaining: 402ms
983:	learn: 0.5492722	total: 23.3s	remaining: 378ms
984:	learn: 0.5491619	total: 23.3s	remaining: 355ms
985:	learn: 0.5491324	total: 23.3s	remaining: 331ms
986:	learn: 0.5488963	total: 23.3s	remaining: 307ms
987:	learn: 0.5488793	total: 23.4s	remaining: 284ms
988:	learn: 0.5488297	total: 23.4s	remaining: 260ms
989:	learn: 0.5485191	total: 23.4s	remaining: 236ms
990:	learn: 0.5484840	total: 23.4s	remaining: 213ms
991:	learn: 0.5483895	total: 23.4s	remaining: 189ms
992:	learn: 0.5483503	total: 23.5s	remaining: 165ms
993:	learn: 

140:	learn: 0.9126669	total: 3.89s	remaining: 23.7s
141:	learn: 0.9107561	total: 3.91s	remaining: 23.7s
142:	learn: 0.9088480	total: 3.94s	remaining: 23.6s
143:	learn: 0.9062251	total: 3.97s	remaining: 23.6s
144:	learn: 0.9038425	total: 4s	remaining: 23.6s
145:	learn: 0.9021034	total: 4.03s	remaining: 23.5s
146:	learn: 0.8997283	total: 4.05s	remaining: 23.5s
147:	learn: 0.8989581	total: 4.08s	remaining: 23.5s
148:	learn: 0.8970804	total: 4.1s	remaining: 23.4s
149:	learn: 0.8955309	total: 4.13s	remaining: 23.4s
150:	learn: 0.8934046	total: 4.16s	remaining: 23.4s
151:	learn: 0.8909071	total: 4.19s	remaining: 23.4s
152:	learn: 0.8898060	total: 4.21s	remaining: 23.3s
153:	learn: 0.8883496	total: 4.23s	remaining: 23.3s
154:	learn: 0.8868418	total: 4.26s	remaining: 23.2s
155:	learn: 0.8849997	total: 4.28s	remaining: 23.2s
156:	learn: 0.8836242	total: 4.31s	remaining: 23.1s
157:	learn: 0.8821186	total: 4.33s	remaining: 23.1s
158:	learn: 0.8805766	total: 4.35s	remaining: 23s
159:	learn: 0.8787

303:	learn: 0.7322410	total: 7.98s	remaining: 18.3s
304:	learn: 0.7318871	total: 8s	remaining: 18.2s
305:	learn: 0.7310256	total: 8.03s	remaining: 18.2s
306:	learn: 0.7305161	total: 8.05s	remaining: 18.2s
307:	learn: 0.7295092	total: 8.08s	remaining: 18.1s
308:	learn: 0.7286617	total: 8.1s	remaining: 18.1s
309:	learn: 0.7281653	total: 8.12s	remaining: 18.1s
310:	learn: 0.7274784	total: 8.15s	remaining: 18.1s
311:	learn: 0.7263550	total: 8.18s	remaining: 18s
312:	learn: 0.7257946	total: 8.2s	remaining: 18s
313:	learn: 0.7253870	total: 8.22s	remaining: 18s
314:	learn: 0.7246342	total: 8.25s	remaining: 17.9s
315:	learn: 0.7241323	total: 8.28s	remaining: 17.9s
316:	learn: 0.7231983	total: 8.3s	remaining: 17.9s
317:	learn: 0.7223153	total: 8.34s	remaining: 17.9s
318:	learn: 0.7218750	total: 8.36s	remaining: 17.8s
319:	learn: 0.7216320	total: 8.38s	remaining: 17.8s
320:	learn: 0.7209555	total: 8.41s	remaining: 17.8s
321:	learn: 0.7198230	total: 8.43s	remaining: 17.8s
322:	learn: 0.7190146	to

467:	learn: 0.6617120	total: 11.8s	remaining: 13.4s
468:	learn: 0.6615058	total: 11.8s	remaining: 13.4s
469:	learn: 0.6613107	total: 11.8s	remaining: 13.3s
470:	learn: 0.6608542	total: 11.8s	remaining: 13.3s
471:	learn: 0.6605810	total: 11.9s	remaining: 13.3s
472:	learn: 0.6602705	total: 11.9s	remaining: 13.2s
473:	learn: 0.6601248	total: 11.9s	remaining: 13.2s
474:	learn: 0.6597615	total: 11.9s	remaining: 13.2s
475:	learn: 0.6590883	total: 12s	remaining: 13.2s
476:	learn: 0.6586483	total: 12s	remaining: 13.1s
477:	learn: 0.6584657	total: 12s	remaining: 13.1s
478:	learn: 0.6579310	total: 12s	remaining: 13.1s
479:	learn: 0.6575326	total: 12s	remaining: 13s
480:	learn: 0.6570793	total: 12.1s	remaining: 13s
481:	learn: 0.6569663	total: 12.1s	remaining: 13s
482:	learn: 0.6566529	total: 12.1s	remaining: 13s
483:	learn: 0.6562534	total: 12.1s	remaining: 12.9s
484:	learn: 0.6555299	total: 12.2s	remaining: 12.9s
485:	learn: 0.6552280	total: 12.2s	remaining: 12.9s
486:	learn: 0.6547867	total: 1

629:	learn: 0.6190401	total: 15.4s	remaining: 9.02s
630:	learn: 0.6187031	total: 15.4s	remaining: 9s
631:	learn: 0.6184402	total: 15.4s	remaining: 8.97s
632:	learn: 0.6182507	total: 15.4s	remaining: 8.95s
633:	learn: 0.6182095	total: 15.5s	remaining: 8.92s
634:	learn: 0.6180045	total: 15.5s	remaining: 8.9s
635:	learn: 0.6179493	total: 15.5s	remaining: 8.87s
636:	learn: 0.6177882	total: 15.5s	remaining: 8.84s
637:	learn: 0.6174905	total: 15.5s	remaining: 8.81s
638:	learn: 0.6173739	total: 15.6s	remaining: 8.79s
639:	learn: 0.6172997	total: 15.6s	remaining: 8.76s
640:	learn: 0.6168997	total: 15.6s	remaining: 8.74s
641:	learn: 0.6166271	total: 15.6s	remaining: 8.71s
642:	learn: 0.6161062	total: 15.6s	remaining: 8.69s
643:	learn: 0.6158222	total: 15.7s	remaining: 8.66s
644:	learn: 0.6156766	total: 15.7s	remaining: 8.63s
645:	learn: 0.6149672	total: 15.7s	remaining: 8.61s
646:	learn: 0.6149370	total: 15.7s	remaining: 8.59s
647:	learn: 0.6148663	total: 15.8s	remaining: 8.56s
648:	learn: 0.61

791:	learn: 0.5811972	total: 19s	remaining: 4.98s
792:	learn: 0.5811076	total: 19s	remaining: 4.96s
793:	learn: 0.5807945	total: 19s	remaining: 4.93s
794:	learn: 0.5805538	total: 19s	remaining: 4.91s
795:	learn: 0.5803551	total: 19s	remaining: 4.88s
796:	learn: 0.5801922	total: 19.1s	remaining: 4.86s
797:	learn: 0.5799783	total: 19.1s	remaining: 4.83s
798:	learn: 0.5797883	total: 19.1s	remaining: 4.81s
799:	learn: 0.5796797	total: 19.1s	remaining: 4.78s
800:	learn: 0.5795138	total: 19.2s	remaining: 4.76s
801:	learn: 0.5793826	total: 19.2s	remaining: 4.73s
802:	learn: 0.5793271	total: 19.2s	remaining: 4.71s
803:	learn: 0.5792738	total: 19.2s	remaining: 4.68s
804:	learn: 0.5789441	total: 19.2s	remaining: 4.66s
805:	learn: 0.5788587	total: 19.3s	remaining: 4.64s
806:	learn: 0.5787866	total: 19.3s	remaining: 4.61s
807:	learn: 0.5783232	total: 19.3s	remaining: 4.59s
808:	learn: 0.5781607	total: 19.3s	remaining: 4.57s
809:	learn: 0.5779312	total: 19.4s	remaining: 4.54s
810:	learn: 0.5775984	

954:	learn: 0.5545910	total: 22.5s	remaining: 1.06s
955:	learn: 0.5544152	total: 22.5s	remaining: 1.04s
956:	learn: 0.5542962	total: 22.6s	remaining: 1.01s
957:	learn: 0.5542576	total: 22.6s	remaining: 990ms
958:	learn: 0.5541297	total: 22.6s	remaining: 966ms
959:	learn: 0.5538657	total: 22.6s	remaining: 943ms
960:	learn: 0.5537298	total: 22.6s	remaining: 919ms
961:	learn: 0.5536959	total: 22.7s	remaining: 895ms
962:	learn: 0.5535759	total: 22.7s	remaining: 872ms
963:	learn: 0.5535064	total: 22.7s	remaining: 848ms
964:	learn: 0.5534230	total: 22.7s	remaining: 824ms
965:	learn: 0.5533927	total: 22.8s	remaining: 801ms
966:	learn: 0.5531439	total: 22.8s	remaining: 777ms
967:	learn: 0.5529666	total: 22.8s	remaining: 753ms
968:	learn: 0.5529605	total: 22.8s	remaining: 730ms
969:	learn: 0.5528705	total: 22.8s	remaining: 706ms
970:	learn: 0.5526179	total: 22.9s	remaining: 683ms
971:	learn: 0.5524785	total: 22.9s	remaining: 659ms
972:	learn: 0.5520479	total: 22.9s	remaining: 636ms
973:	learn: 

119:	learn: 0.9448076	total: 3.4s	remaining: 24.9s
120:	learn: 0.9420956	total: 3.42s	remaining: 24.9s
121:	learn: 0.9394791	total: 3.46s	remaining: 24.9s
122:	learn: 0.9380543	total: 3.48s	remaining: 24.8s
123:	learn: 0.9365684	total: 3.5s	remaining: 24.7s
124:	learn: 0.9337292	total: 3.54s	remaining: 24.8s
125:	learn: 0.9312992	total: 3.57s	remaining: 24.8s
126:	learn: 0.9287102	total: 3.6s	remaining: 24.7s
127:	learn: 0.9253932	total: 3.63s	remaining: 24.7s
128:	learn: 0.9236906	total: 3.65s	remaining: 24.7s
129:	learn: 0.9217749	total: 3.68s	remaining: 24.6s
130:	learn: 0.9206640	total: 3.7s	remaining: 24.6s
131:	learn: 0.9185558	total: 3.73s	remaining: 24.5s
132:	learn: 0.9168091	total: 3.76s	remaining: 24.5s
133:	learn: 0.9153075	total: 3.78s	remaining: 24.4s
134:	learn: 0.9132352	total: 3.81s	remaining: 24.4s
135:	learn: 0.9118063	total: 3.83s	remaining: 24.3s
136:	learn: 0.9100429	total: 3.86s	remaining: 24.3s
137:	learn: 0.9074744	total: 3.88s	remaining: 24.3s
138:	learn: 0.90

279:	learn: 0.7371686	total: 7.49s	remaining: 19.3s
280:	learn: 0.7361012	total: 7.52s	remaining: 19.2s
281:	learn: 0.7359119	total: 7.54s	remaining: 19.2s
282:	learn: 0.7352544	total: 7.57s	remaining: 19.2s
283:	learn: 0.7346038	total: 7.59s	remaining: 19.1s
284:	learn: 0.7336639	total: 7.62s	remaining: 19.1s
285:	learn: 0.7330876	total: 7.64s	remaining: 19.1s
286:	learn: 0.7329141	total: 7.66s	remaining: 19s
287:	learn: 0.7326787	total: 7.68s	remaining: 19s
288:	learn: 0.7321147	total: 7.71s	remaining: 19s
289:	learn: 0.7313678	total: 7.74s	remaining: 18.9s
290:	learn: 0.7308118	total: 7.76s	remaining: 18.9s
291:	learn: 0.7302635	total: 7.78s	remaining: 18.9s
292:	learn: 0.7298098	total: 7.8s	remaining: 18.8s
293:	learn: 0.7291184	total: 7.83s	remaining: 18.8s
294:	learn: 0.7283920	total: 7.85s	remaining: 18.8s
295:	learn: 0.7276549	total: 7.88s	remaining: 18.7s
296:	learn: 0.7270009	total: 7.9s	remaining: 18.7s
297:	learn: 0.7265151	total: 7.92s	remaining: 18.7s
298:	learn: 0.725281

440:	learn: 0.6584963	total: 11.3s	remaining: 14.3s
441:	learn: 0.6584585	total: 11.3s	remaining: 14.3s
442:	learn: 0.6584306	total: 11.3s	remaining: 14.2s
443:	learn: 0.6582694	total: 11.3s	remaining: 14.2s
444:	learn: 0.6580490	total: 11.4s	remaining: 14.2s
445:	learn: 0.6577316	total: 11.4s	remaining: 14.2s
446:	learn: 0.6575374	total: 11.4s	remaining: 14.1s
447:	learn: 0.6574921	total: 11.4s	remaining: 14.1s
448:	learn: 0.6572745	total: 11.5s	remaining: 14.1s
449:	learn: 0.6571585	total: 11.5s	remaining: 14s
450:	learn: 0.6570639	total: 11.5s	remaining: 14s
451:	learn: 0.6568527	total: 11.5s	remaining: 14s
452:	learn: 0.6565534	total: 11.5s	remaining: 13.9s
453:	learn: 0.6565365	total: 11.6s	remaining: 13.9s
454:	learn: 0.6562069	total: 11.6s	remaining: 13.9s
455:	learn: 0.6559376	total: 11.6s	remaining: 13.8s
456:	learn: 0.6554188	total: 11.6s	remaining: 13.8s
457:	learn: 0.6553093	total: 11.6s	remaining: 13.8s
458:	learn: 0.6545743	total: 11.7s	remaining: 13.7s
459:	learn: 0.6542

603:	learn: 0.6151623	total: 14.9s	remaining: 9.78s
604:	learn: 0.6148268	total: 14.9s	remaining: 9.75s
605:	learn: 0.6145504	total: 15s	remaining: 9.72s
606:	learn: 0.6142917	total: 15s	remaining: 9.7s
607:	learn: 0.6141713	total: 15s	remaining: 9.67s
608:	learn: 0.6141626	total: 15s	remaining: 9.64s
609:	learn: 0.6133114	total: 15.1s	remaining: 9.62s
610:	learn: 0.6132552	total: 15.1s	remaining: 9.59s
611:	learn: 0.6131347	total: 15.1s	remaining: 9.57s
612:	learn: 0.6130874	total: 15.1s	remaining: 9.54s
613:	learn: 0.6130747	total: 15.1s	remaining: 9.51s
614:	learn: 0.6130111	total: 15.2s	remaining: 9.48s
615:	learn: 0.6126692	total: 15.2s	remaining: 9.46s
616:	learn: 0.6125000	total: 15.2s	remaining: 9.43s
617:	learn: 0.6121267	total: 15.2s	remaining: 9.41s
618:	learn: 0.6116797	total: 15.2s	remaining: 9.38s
619:	learn: 0.6116292	total: 15.3s	remaining: 9.36s
620:	learn: 0.6115250	total: 15.3s	remaining: 9.33s
621:	learn: 0.6115149	total: 15.3s	remaining: 9.3s
622:	learn: 0.6114686	

770:	learn: 0.5821544	total: 18.5s	remaining: 5.51s
771:	learn: 0.5820211	total: 18.6s	remaining: 5.48s
772:	learn: 0.5816301	total: 18.6s	remaining: 5.46s
773:	learn: 0.5812483	total: 18.6s	remaining: 5.44s
774:	learn: 0.5810365	total: 18.6s	remaining: 5.41s
775:	learn: 0.5808007	total: 18.7s	remaining: 5.39s
776:	learn: 0.5807878	total: 18.7s	remaining: 5.36s
777:	learn: 0.5804669	total: 18.7s	remaining: 5.34s
778:	learn: 0.5803853	total: 18.7s	remaining: 5.31s
779:	learn: 0.5802968	total: 18.7s	remaining: 5.29s
780:	learn: 0.5798488	total: 18.8s	remaining: 5.26s
781:	learn: 0.5796555	total: 18.8s	remaining: 5.24s
782:	learn: 0.5795367	total: 18.8s	remaining: 5.21s
783:	learn: 0.5792865	total: 18.8s	remaining: 5.19s
784:	learn: 0.5790861	total: 18.9s	remaining: 5.17s
785:	learn: 0.5789928	total: 18.9s	remaining: 5.14s
786:	learn: 0.5787241	total: 18.9s	remaining: 5.12s
787:	learn: 0.5785962	total: 18.9s	remaining: 5.09s
788:	learn: 0.5782436	total: 18.9s	remaining: 5.07s
789:	learn: 

929:	learn: 0.5581627	total: 21.9s	remaining: 1.65s
930:	learn: 0.5580574	total: 22s	remaining: 1.63s
931:	learn: 0.5579018	total: 22s	remaining: 1.6s
932:	learn: 0.5578039	total: 22s	remaining: 1.58s
933:	learn: 0.5575523	total: 22s	remaining: 1.56s
934:	learn: 0.5573198	total: 22.1s	remaining: 1.53s
935:	learn: 0.5571354	total: 22.1s	remaining: 1.51s
936:	learn: 0.5570686	total: 22.1s	remaining: 1.49s
937:	learn: 0.5568895	total: 22.1s	remaining: 1.46s
938:	learn: 0.5565913	total: 22.2s	remaining: 1.44s
939:	learn: 0.5564034	total: 22.2s	remaining: 1.42s
940:	learn: 0.5560889	total: 22.2s	remaining: 1.39s
941:	learn: 0.5557714	total: 22.2s	remaining: 1.37s
942:	learn: 0.5557528	total: 22.2s	remaining: 1.34s
943:	learn: 0.5555966	total: 22.3s	remaining: 1.32s
944:	learn: 0.5553402	total: 22.3s	remaining: 1.3s
945:	learn: 0.5551829	total: 22.3s	remaining: 1.27s
946:	learn: 0.5551283	total: 22.3s	remaining: 1.25s
947:	learn: 0.5549413	total: 22.4s	remaining: 1.23s
948:	learn: 0.5547304	

90:	learn: 1.0293356	total: 2.6s	remaining: 26s
91:	learn: 1.0264452	total: 2.63s	remaining: 26s
92:	learn: 1.0233243	total: 2.66s	remaining: 25.9s
93:	learn: 1.0189747	total: 2.69s	remaining: 25.9s
94:	learn: 1.0154992	total: 2.72s	remaining: 25.9s
95:	learn: 1.0126491	total: 2.75s	remaining: 25.9s
96:	learn: 1.0099274	total: 2.77s	remaining: 25.8s
97:	learn: 1.0071596	total: 2.8s	remaining: 25.8s
98:	learn: 1.0045269	total: 2.82s	remaining: 25.7s
99:	learn: 0.9998786	total: 2.85s	remaining: 25.7s
100:	learn: 0.9974033	total: 2.88s	remaining: 25.7s
101:	learn: 0.9950917	total: 2.91s	remaining: 25.6s
102:	learn: 0.9922172	total: 2.93s	remaining: 25.5s
103:	learn: 0.9899741	total: 2.96s	remaining: 25.5s
104:	learn: 0.9884914	total: 2.98s	remaining: 25.4s
105:	learn: 0.9863766	total: 3s	remaining: 25.3s
106:	learn: 0.9846024	total: 3.03s	remaining: 25.3s
107:	learn: 0.9829340	total: 3.05s	remaining: 25.2s
108:	learn: 0.9791258	total: 3.08s	remaining: 25.2s
109:	learn: 0.9757690	total: 3.

250:	learn: 0.7623464	total: 6.81s	remaining: 20.3s
251:	learn: 0.7611397	total: 6.83s	remaining: 20.3s
252:	learn: 0.7606156	total: 6.85s	remaining: 20.2s
253:	learn: 0.7597658	total: 6.88s	remaining: 20.2s
254:	learn: 0.7593101	total: 6.9s	remaining: 20.2s
255:	learn: 0.7589562	total: 6.92s	remaining: 20.1s
256:	learn: 0.7581350	total: 6.95s	remaining: 20.1s
257:	learn: 0.7571319	total: 6.98s	remaining: 20.1s
258:	learn: 0.7563497	total: 7s	remaining: 20s
259:	learn: 0.7554592	total: 7.03s	remaining: 20s
260:	learn: 0.7546078	total: 7.06s	remaining: 20s
261:	learn: 0.7540674	total: 7.08s	remaining: 19.9s
262:	learn: 0.7536282	total: 7.1s	remaining: 19.9s
263:	learn: 0.7529689	total: 7.12s	remaining: 19.9s
264:	learn: 0.7524225	total: 7.14s	remaining: 19.8s
265:	learn: 0.7513781	total: 7.17s	remaining: 19.8s
266:	learn: 0.7495364	total: 7.21s	remaining: 19.8s
267:	learn: 0.7489778	total: 7.23s	remaining: 19.7s
268:	learn: 0.7476502	total: 7.26s	remaining: 19.7s
269:	learn: 0.7474323	t

409:	learn: 0.6756298	total: 10.6s	remaining: 15.2s
410:	learn: 0.6755191	total: 10.6s	remaining: 15.2s
411:	learn: 0.6751254	total: 10.6s	remaining: 15.2s
412:	learn: 0.6748161	total: 10.6s	remaining: 15.1s
413:	learn: 0.6744377	total: 10.7s	remaining: 15.1s
414:	learn: 0.6744202	total: 10.7s	remaining: 15.1s
415:	learn: 0.6743795	total: 10.7s	remaining: 15s
416:	learn: 0.6736812	total: 10.7s	remaining: 15s
417:	learn: 0.6730813	total: 10.8s	remaining: 15s
418:	learn: 0.6720205	total: 10.8s	remaining: 15s
419:	learn: 0.6715195	total: 10.8s	remaining: 14.9s
420:	learn: 0.6712925	total: 10.8s	remaining: 14.9s
421:	learn: 0.6712693	total: 10.9s	remaining: 14.9s
422:	learn: 0.6706206	total: 10.9s	remaining: 14.8s
423:	learn: 0.6701296	total: 10.9s	remaining: 14.8s
424:	learn: 0.6700021	total: 10.9s	remaining: 14.8s
425:	learn: 0.6693558	total: 10.9s	remaining: 14.7s
426:	learn: 0.6687982	total: 11s	remaining: 14.7s
427:	learn: 0.6683679	total: 11s	remaining: 14.7s
428:	learn: 0.6681740	to

568:	learn: 0.6275676	total: 14.1s	remaining: 10.7s
569:	learn: 0.6273481	total: 14.2s	remaining: 10.7s
570:	learn: 0.6270856	total: 14.2s	remaining: 10.7s
571:	learn: 0.6270719	total: 14.2s	remaining: 10.6s
572:	learn: 0.6270046	total: 14.2s	remaining: 10.6s
573:	learn: 0.6265694	total: 14.3s	remaining: 10.6s
574:	learn: 0.6264097	total: 14.3s	remaining: 10.5s
575:	learn: 0.6258819	total: 14.3s	remaining: 10.5s
576:	learn: 0.6257607	total: 14.3s	remaining: 10.5s
577:	learn: 0.6255019	total: 14.3s	remaining: 10.5s
578:	learn: 0.6252573	total: 14.4s	remaining: 10.4s
579:	learn: 0.6248232	total: 14.4s	remaining: 10.4s
580:	learn: 0.6245255	total: 14.4s	remaining: 10.4s
581:	learn: 0.6243767	total: 14.4s	remaining: 10.4s
582:	learn: 0.6241289	total: 14.5s	remaining: 10.3s
583:	learn: 0.6240459	total: 14.5s	remaining: 10.3s
584:	learn: 0.6238342	total: 14.5s	remaining: 10.3s
585:	learn: 0.6236022	total: 14.5s	remaining: 10.3s
586:	learn: 0.6233020	total: 14.5s	remaining: 10.2s
587:	learn: 

731:	learn: 0.5904330	total: 17.8s	remaining: 6.5s
732:	learn: 0.5899330	total: 17.8s	remaining: 6.48s
733:	learn: 0.5897400	total: 17.8s	remaining: 6.45s
734:	learn: 0.5895882	total: 17.8s	remaining: 6.43s
735:	learn: 0.5893090	total: 17.8s	remaining: 6.4s
736:	learn: 0.5891137	total: 17.9s	remaining: 6.38s
737:	learn: 0.5890935	total: 17.9s	remaining: 6.35s
738:	learn: 0.5889781	total: 17.9s	remaining: 6.33s
739:	learn: 0.5888003	total: 17.9s	remaining: 6.3s
740:	learn: 0.5885534	total: 18s	remaining: 6.28s
741:	learn: 0.5883800	total: 18s	remaining: 6.25s
742:	learn: 0.5882026	total: 18s	remaining: 6.23s
743:	learn: 0.5881876	total: 18s	remaining: 6.2s
744:	learn: 0.5876034	total: 18.1s	remaining: 6.18s
745:	learn: 0.5873548	total: 18.1s	remaining: 6.16s
746:	learn: 0.5871510	total: 18.1s	remaining: 6.13s
747:	learn: 0.5868375	total: 18.1s	remaining: 6.11s
748:	learn: 0.5865842	total: 18.2s	remaining: 6.08s
749:	learn: 0.5863315	total: 18.2s	remaining: 6.06s
750:	learn: 0.5861062	to

892:	learn: 0.5611210	total: 21.3s	remaining: 2.56s
893:	learn: 0.5609972	total: 21.4s	remaining: 2.53s
894:	learn: 0.5608443	total: 21.4s	remaining: 2.51s
895:	learn: 0.5606633	total: 21.4s	remaining: 2.48s
896:	learn: 0.5606223	total: 21.4s	remaining: 2.46s
897:	learn: 0.5604269	total: 21.4s	remaining: 2.44s
898:	learn: 0.5602952	total: 21.5s	remaining: 2.41s
899:	learn: 0.5601422	total: 21.5s	remaining: 2.39s
900:	learn: 0.5600532	total: 21.5s	remaining: 2.36s
901:	learn: 0.5600467	total: 21.5s	remaining: 2.34s
902:	learn: 0.5598955	total: 21.5s	remaining: 2.31s
903:	learn: 0.5598246	total: 21.6s	remaining: 2.29s
904:	learn: 0.5596080	total: 21.6s	remaining: 2.27s
905:	learn: 0.5594784	total: 21.6s	remaining: 2.24s
906:	learn: 0.5592904	total: 21.6s	remaining: 2.22s
907:	learn: 0.5591274	total: 21.7s	remaining: 2.19s
908:	learn: 0.5587115	total: 21.7s	remaining: 2.17s
909:	learn: 0.5586090	total: 21.7s	remaining: 2.15s
910:	learn: 0.5584690	total: 21.7s	remaining: 2.12s
911:	learn: 

56:	learn: 1.1798100	total: 1.71s	remaining: 28.3s
57:	learn: 1.1739989	total: 1.74s	remaining: 28.3s
58:	learn: 1.1664420	total: 1.77s	remaining: 28.2s
59:	learn: 1.1622708	total: 1.79s	remaining: 28.1s
60:	learn: 1.1569000	total: 1.82s	remaining: 28s
61:	learn: 1.1540067	total: 1.84s	remaining: 27.9s
62:	learn: 1.1472627	total: 1.88s	remaining: 27.9s
63:	learn: 1.1420035	total: 1.91s	remaining: 27.9s
64:	learn: 1.1391409	total: 1.93s	remaining: 27.8s
65:	learn: 1.1353714	total: 1.95s	remaining: 27.6s
66:	learn: 1.1300382	total: 1.98s	remaining: 27.6s
67:	learn: 1.1220674	total: 2.02s	remaining: 27.6s
68:	learn: 1.1162820	total: 2.05s	remaining: 27.6s
69:	learn: 1.1113046	total: 2.08s	remaining: 27.6s
70:	learn: 1.1073931	total: 2.1s	remaining: 27.5s
71:	learn: 1.1035705	total: 2.13s	remaining: 27.4s
72:	learn: 1.0990490	total: 2.15s	remaining: 27.4s
73:	learn: 1.0952577	total: 2.19s	remaining: 27.4s
74:	learn: 1.0911518	total: 2.21s	remaining: 27.3s
75:	learn: 1.0875511	total: 2.24s	

223:	learn: 0.7931231	total: 6.15s	remaining: 21.3s
224:	learn: 0.7913881	total: 6.18s	remaining: 21.3s
225:	learn: 0.7906723	total: 6.21s	remaining: 21.3s
226:	learn: 0.7891138	total: 6.23s	remaining: 21.2s
227:	learn: 0.7883581	total: 6.25s	remaining: 21.2s
228:	learn: 0.7875742	total: 6.28s	remaining: 21.1s
229:	learn: 0.7866106	total: 6.3s	remaining: 21.1s
230:	learn: 0.7858661	total: 6.33s	remaining: 21.1s
231:	learn: 0.7846723	total: 6.35s	remaining: 21s
232:	learn: 0.7838825	total: 6.38s	remaining: 21s
233:	learn: 0.7826410	total: 6.41s	remaining: 21s
234:	learn: 0.7814641	total: 6.44s	remaining: 21s
235:	learn: 0.7806094	total: 6.46s	remaining: 20.9s
236:	learn: 0.7797374	total: 6.49s	remaining: 20.9s
237:	learn: 0.7791184	total: 6.51s	remaining: 20.9s
238:	learn: 0.7783601	total: 6.54s	remaining: 20.8s
239:	learn: 0.7773395	total: 6.57s	remaining: 20.8s
240:	learn: 0.7764645	total: 6.59s	remaining: 20.8s
241:	learn: 0.7759998	total: 6.62s	remaining: 20.7s
242:	learn: 0.7750616

388:	learn: 0.6863013	total: 10.2s	remaining: 16s
389:	learn: 0.6861551	total: 10.2s	remaining: 15.9s
390:	learn: 0.6858912	total: 10.2s	remaining: 15.9s
391:	learn: 0.6856086	total: 10.2s	remaining: 15.9s
392:	learn: 0.6854120	total: 10.3s	remaining: 15.8s
393:	learn: 0.6847889	total: 10.3s	remaining: 15.8s
394:	learn: 0.6837142	total: 10.3s	remaining: 15.8s
395:	learn: 0.6836748	total: 10.3s	remaining: 15.8s
396:	learn: 0.6833899	total: 10.3s	remaining: 15.7s
397:	learn: 0.6829273	total: 10.4s	remaining: 15.7s
398:	learn: 0.6824194	total: 10.4s	remaining: 15.7s
399:	learn: 0.6814896	total: 10.4s	remaining: 15.6s
400:	learn: 0.6811080	total: 10.5s	remaining: 15.6s
401:	learn: 0.6804587	total: 10.5s	remaining: 15.6s
402:	learn: 0.6803057	total: 10.5s	remaining: 15.6s
403:	learn: 0.6799028	total: 10.5s	remaining: 15.5s
404:	learn: 0.6796116	total: 10.5s	remaining: 15.5s
405:	learn: 0.6787830	total: 10.6s	remaining: 15.5s
406:	learn: 0.6784838	total: 10.6s	remaining: 15.4s
407:	learn: 0.

555:	learn: 0.6282910	total: 14s	remaining: 11.2s
556:	learn: 0.6280803	total: 14s	remaining: 11.1s
557:	learn: 0.6278021	total: 14s	remaining: 11.1s
558:	learn: 0.6270438	total: 14.1s	remaining: 11.1s
559:	learn: 0.6266134	total: 14.1s	remaining: 11.1s
560:	learn: 0.6262188	total: 14.1s	remaining: 11s
561:	learn: 0.6259295	total: 14.1s	remaining: 11s
562:	learn: 0.6255861	total: 14.2s	remaining: 11s
563:	learn: 0.6252716	total: 14.2s	remaining: 11s
564:	learn: 0.6252021	total: 14.2s	remaining: 10.9s
565:	learn: 0.6248314	total: 14.2s	remaining: 10.9s
566:	learn: 0.6243077	total: 14.3s	remaining: 10.9s
567:	learn: 0.6239913	total: 14.3s	remaining: 10.9s
568:	learn: 0.6236735	total: 14.3s	remaining: 10.8s
569:	learn: 0.6233779	total: 14.3s	remaining: 10.8s
570:	learn: 0.6231446	total: 14.4s	remaining: 10.8s
571:	learn: 0.6231295	total: 14.4s	remaining: 10.8s
572:	learn: 0.6228158	total: 14.4s	remaining: 10.7s
573:	learn: 0.6223114	total: 14.4s	remaining: 10.7s
574:	learn: 0.6220544	tota

721:	learn: 0.5888629	total: 17.7s	remaining: 6.83s
722:	learn: 0.5884379	total: 17.8s	remaining: 6.8s
723:	learn: 0.5883211	total: 17.8s	remaining: 6.78s
724:	learn: 0.5882151	total: 17.8s	remaining: 6.75s
725:	learn: 0.5882051	total: 17.8s	remaining: 6.72s
726:	learn: 0.5881339	total: 17.8s	remaining: 6.7s
727:	learn: 0.5880096	total: 17.9s	remaining: 6.67s
728:	learn: 0.5878065	total: 17.9s	remaining: 6.65s
729:	learn: 0.5874371	total: 17.9s	remaining: 6.62s
730:	learn: 0.5871852	total: 17.9s	remaining: 6.6s
731:	learn: 0.5870581	total: 18s	remaining: 6.57s
732:	learn: 0.5868151	total: 18s	remaining: 6.55s
733:	learn: 0.5864357	total: 18s	remaining: 6.52s
734:	learn: 0.5862752	total: 18s	remaining: 6.5s
735:	learn: 0.5859603	total: 18s	remaining: 6.47s
736:	learn: 0.5858518	total: 18.1s	remaining: 6.45s
737:	learn: 0.5856763	total: 18.1s	remaining: 6.42s
738:	learn: 0.5855765	total: 18.1s	remaining: 6.39s
739:	learn: 0.5854832	total: 18.1s	remaining: 6.37s
740:	learn: 0.5852849	tota

888:	learn: 0.5580715	total: 21.4s	remaining: 2.68s
889:	learn: 0.5578022	total: 21.5s	remaining: 2.65s
890:	learn: 0.5576568	total: 21.5s	remaining: 2.63s
891:	learn: 0.5573773	total: 21.5s	remaining: 2.6s
892:	learn: 0.5573587	total: 21.5s	remaining: 2.58s
893:	learn: 0.5573095	total: 21.5s	remaining: 2.56s
894:	learn: 0.5572777	total: 21.6s	remaining: 2.53s
895:	learn: 0.5571358	total: 21.6s	remaining: 2.5s
896:	learn: 0.5569426	total: 21.6s	remaining: 2.48s
897:	learn: 0.5567775	total: 21.6s	remaining: 2.46s
898:	learn: 0.5567247	total: 21.7s	remaining: 2.43s
899:	learn: 0.5564138	total: 21.7s	remaining: 2.41s
900:	learn: 0.5563403	total: 21.7s	remaining: 2.38s
901:	learn: 0.5562658	total: 21.7s	remaining: 2.36s
902:	learn: 0.5559279	total: 21.7s	remaining: 2.33s
903:	learn: 0.5557405	total: 21.8s	remaining: 2.31s
904:	learn: 0.5556693	total: 21.8s	remaining: 2.29s
905:	learn: 0.5554472	total: 21.8s	remaining: 2.26s
906:	learn: 0.5551573	total: 21.8s	remaining: 2.24s
907:	learn: 0.

0.7861668990156632

Cross validation score for ngrams with letters + some symbols: 0.7858903433902703

Cross validation score for ngrams with letters only: 0.7861668990156632